### PR TITLE
SDAP-36 MUDROD Storage unit test suite

### PR DESCRIPTION
**Pull request by Yongyao**
@lewismc We did some initial development. Could you please take a look and let us know what you think? I saw the discussion on JIRA. Do you think we should use JUnit to do all the testing or implement a generic testing class like gora: https://github.com/apache/gora/blob/master/gora-solr/src/test/java/org/apache/gora/solr/GoraSolrTestDriver.java ?

### CODE REVIEW

**Comment by asfgit** [comment]
Can one of the admins verify this patch?

File: core/src/test/java/org/apache/sdap/mudrod/driver/ESDriverTest.java [hunk 1]
```diff
@@ -0,0 +1,394 @@
+package org.apache.sdap.mudrod.driver;
```
Comment (lewismc) [inline core/src/test/java/org/apache/sdap/mudrod/driver/ESDriverTest.java:14]: Please add license header

File: core/src/test/java/org/apache/sdap/mudrod/driver/ESDriverTest.java [hunk 2]
```diff
@@ -0,0 +1,394 @@
...
+
+import static org.junit.Assert.*;
```
Comment (lewismc) [inline core/src/test/java/org/apache/sdap/mudrod/driver/ESDriverTest.java]: Never use wildcard imports always make explicit imports

**Review by lewismc — CHANGES_REQUESTED** [review]
@Yongyao this looks like a decent start, the only issue however is that you are actually not testing anything!!!
Please have a look at the JUnit tutorial and you will see how to make use of **assert** in your tests

https://github.com/junit-team/junit4/wiki/Getting-started

File: core/src/test/java/org/apache/sdap/mudrod/driver/ESDriverTest.java [hunk 3]
```diff
@@ -28,367 +42,331 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-public class ESDriverTest {
+import junit.framework.Assert;
```
Comment (lewismc) [inline core/src/test/java/org/apache/sdap/mudrod/driver/ESDriverTest.java]: This is the old JUnit import. Please use org.junit....

File: core/src/test/java/org/apache/sdap/mudrod/driver/ESDriverTest.java [hunk 4]
```diff
@@ -11,6 +22,8 @@
 import java.util.Properties;
 import java.util.concurrent.ExecutionException;
 
+import javax.print.DocFlavor.STRING;
```
Comment (lewismc) [inline core/src/test/java/org/apache/sdap/mudrod/driver/ESDriverTest.java]: Remove this import.

**Review by lewismc — CHANGES_REQUESTED** [review]
@clarkgogogo MUDROD source code is formatted with 2 space indents. Please update the PR to accomodate that. Thanks

**Comment by Yongyao** [comment]
@lewismc I changed the format issue before submitting it. Are you looking at the previous version?

File: core/src/test/java/org/apache/sdap/mudrod/driver/ESDriverTest.java [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+
+    BulkProcessor begin = es.getBulkProcessor();
+    es.setBulkProcessor(begin);
+
+    BulkProcessor processor = es.getBulkProcessor();
+    assert processor != null;
+  }
+
+  @Test
+  public void testGenerateUpdateRequestStringStringStringStringObject() {
```
Comment (lewismc) [inline core/src/test/java/org/apache/sdap/mudrod/driver/ESDriverTest.java:305]: Change the name of this unit test

File: core/src/test/java/org/apache/sdap/mudrod/driver/ESDriverTest.java [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+    MudrodEngine mudrodEngine = new MudrodEngine();
+    ESDriver es = new ESDriver(mudrodEngine.loadConfig());
+
+    UpdateRequest res = es.generateUpdateRequest("mudrod", "MetadataLinkage", "id_1", "temp", "string_a");
+    assert res != null;
+  }
+
+  @Test
+  public void testGenerateUpdateRequestStringStringStringMapOfStringObject() {
```
Comment (lewismc) [inline core/src/test/java/org/apache/sdap/mudrod/driver/ESDriverTest.java:314]: Change the name of this unit test

File: core/src/test/java/org/apache/sdap/mudrod/driver/ESDriverTest.java [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+
+    Map<String, Object> result = new HashMap<String, Object>();
+    result.put("temp", "string_a");
+
+    UpdateRequest res = es.generateUpdateRequest("mudrod", "MetadataLinkage", "id_1", result);
+    assert res != null;
+  }
+
+  @Test
+  public void testGetDocCountStringStringArray() {
```
Comment (lewismc) [inline core/src/test/java/org/apache/sdap/mudrod/driver/ESDriverTest.java:326]: Change the name of this unit test

File: core/src/test/java/org/apache/sdap/mudrod/driver/ESDriverTest.java [hunk 8]
```diff
@@ -0,0 +1,357 @@
...
+    ESDriver es = new ESDriver(mudrodEngine.loadConfig());
+
+    String box[] = new String[] { "MetadataLinkage" };
+    int res = es.getDocCount("mudrod", box);
+
+    assert res > 0;
+  }
+
+  @Test
+  public void testGetDocCountStringArrayStringArray() {
```
Comment (lewismc) [inline core/src/test/java/org/apache/sdap/mudrod/driver/ESDriverTest.java:337]: Change the name of this unit test

File: core/src/test/java/org/apache/sdap/mudrod/driver/ESDriverTest.java [hunk 9]
```diff
@@ -0,0 +1,357 @@
...
+    ESDriver es = new ESDriver(mudrodEngine.loadConfig());
+
+    String begin[] = new String[] { "mudrod" };
+    String box[] = new String[] { "MetadataLinkage" };
+    int res = es.getDocCount(begin, box);
+    assert res > 0;
+  }
+
+  @Test
+  public void testGetDocCountStringArrayStringArrayQueryBuilder() {
```
Comment (lewismc) [inline core/src/test/java/org/apache/sdap/mudrod/driver/ESDriverTest.java:348]: Change the name of this unit test

**Review by lewismc — CHANGES_REQUESTED** [review]
Also, I noticed that **nothing** is actually interacting with an embedded, test only Elastuc Search server... hence we are not really testing anything much at all. You should look into establishing an [Embedded Elasticsearch Server for Tests](http://cupofjava.de/blog/2012/11/27/embedded-elasticsearch-server-for-tests/). This way you can use it in the @setUpClass and destroy it in the @tearDownClass

**Comment by Yongyao** [comment]
@lewismc Kejin has added the embedded ES server as you suggested, but he got stuck because the client (https://github.com/Yongyao/incubator-sdap-mudrod/blob/SDAP-36/core/src/test/java/org/apache/sdap/mudrod/driver/ESDriverTest.java#L53) he got through this approach is null. I am actually a little confused with the embedded server approach. Why do you need it? The sample code you provided seems to create a ES client, which is the same as the one we created through ESdriver.

**Comment by lewismc** [comment]
Hi @Yongyao couple of questions here

1. I posted a direct link to an example of creating an [embedded ES server](http://cupofjava.de/blog/2012/11/27/embedded-elasticsearch-server-for-tests/)... this is what you need to create. This will then be used when unit tests are executed... makes sense?
1. Is there any reason that @clarkgogogo is submitting code for a PR via your repository? That would seem to me to be missing the point of us bringing him on board for the summer. What happens when you are gone and no-one is able to manage PR's which come into your repository? It would be appreciated if you could point @clarkgogogo to the [CONTRIBUTING GUIDELINES](https://github.com/apache/incubator-sdap-mudrod/blob/master/CONTRIBUTING.md) such that he can understand and practice the development workflow himself.
1. you state "...but he got stuck because the client he got through this approach is null."... this is because he did not create the embedded ES server as I stated and requested above
1. you state "...I am actually a little confused with the embedded server approach. Why do you need it?" The answer is simple... it is so we can actually test the data persistence functionality within MUDROD. Unless we have a running instance of ES, we are unable to interact with any data... does this make sense?
1. You stated "...The sample code you provided seems to create a ES client, which is the same as the one we created through ESdriver." No this is not correct, if you look at the following code, it clearly creates a ```Node``` object which is the ES data node.
```
    public EmbeddedElasticsearchServer(String dataDirectory) {
        this.dataDirectory = dataDirectory;

        ImmutableSettings.Builder elasticsearchSettings = ImmutableSettings.settingsBuilder()
                .put("http.enabled", "false")
                .put("path.data", dataDirectory);

        node = nodeBuilder()
                .local(true)
                .settings(elasticsearchSettings.build())
                .node();
    }

```
Once we've created the ```Node``` with ```@setUpClass``` then we can begin executing Unit tests against that Node.
1. Final thing for now, for the ESDriverTest you guys have actually not tested anything at all. You are confusing use of the [Java assert](https://docs.oracle.com/javase/7/docs/technotes/guides/language/assert.html) with [JUnit's assert](https://junit.org/junit4/javadoc/latest/org/junit/Assert.html)

**Comment by Yongyao** [comment]
@lewismc Thanks for the explanation. Now I see why we need the embedded server. Sure, I will go over the guideline with him next week.

**Comment by jjacob7734** [comment]
Closing this issue as it is not implemented properly and it does not address the task at hand.

**Pull request closed without merging**